File: e0509_gripper_description/scripts/gazebo_bridge.py
#!/usr/bin/env python3
"""
Gazebo Bridge - 실제 로봇 → Gazebo 동기화

실제 로봇의 joint_states를 받아서 Gazebo 로봇의 controller에 전달합니다.
실제 로봇이 움직이면 Gazebo의 로봇도 동일하게 움직입니다.

사용법:
    ros2 run e0509_gripper_description gazebo_bridge.py \
        --real-ns dsr01 --gazebo-ns gz
"""

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from std_msgs.msg import Float64MultiArray
from builtin_interfaces.msg import Duration


class GazeboBridge(Node):
    """실제 로봇 → Gazebo 동기화 브릿지"""

    # 로봇 팔 조인트 이름 (순서 중요)
    ARM_JOINTS = ['joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6']

    # 그리퍼 조인트 이름
    GRIPPER_JOINTS = ['gripper_rh_r1', 'gripper_rh_r2', 'gripper_rh_l1', 'gripper_rh_l2']

    def __init__(self, real_ns='dsr01', gazebo_ns='gz'):
        super().__init__('gazebo_bridge')

        self.real_ns = real_ns
        self.gazebo_ns = gazebo_ns

        # 현재 조인트 위치 저장
        self.joint_positions = {}

        # 실제 로봇 joint_states 구독
        real_topic = f'/{real_ns}/joint_states'
        self.real_sub = self.create_subscription(
            JointState,
            real_topic,
            self.real_joint_state_callback,
            10
        )

        # Gazebo 로봇 arm trajectory controller 발행
        arm_topic = f'/{gazebo_ns}/joint_trajectory_controller/joint_trajectory'
        self.arm_pub = self.create_publisher(JointTrajectory, arm_topic, 10)

        # Gazebo 로봇 gripper controller 발행
        gripper_topic = f'/{gazebo_ns}/gripper_controller/commands'
        self.gripper_pub = self.create_publisher(Float64MultiArray, gripper_topic, 10)

        # 발행 타이머 (50Hz)
        self.timer = self.create_timer(0.02, self.publish_to_gazebo)

        self.get_logger().info('=' * 60)
        self.get_logger().info('  Gazebo Digital Twin Bridge Started')
        self.get_logger().info('=' * 60)
        self.get_logger().info(f'  Real robot namespace: {real_ns}')
        self.get_logger().info(f'  Gazebo namespace: {gazebo_ns}')
        self.get_logger().info(f'  Subscribed to: {real_topic}')
        self.get_logger().info(f'  Publishing to: {arm_topic}')
        self.get_logger().info(f'  Publishing to: {gripper_topic}')
        self.get_logger().info('=' * 60)

        self.count = 0

    def real_joint_state_callback(self, msg: JointState):
        """실제 로봇의 joint_states를 저장"""
        for i, name in enumerate(msg.name):
            if i < len(msg.position):
                self.joint_positions[name] = msg.position[i]

    def publish_to_gazebo(self):
        """저장된 joint 위치를 Gazebo로 발행"""
        if not self.joint_positions:
            return

        # Arm trajectory 발행
        arm_positions = []
        for joint in self.ARM_JOINTS:
            if joint in self.joint_positions:
                arm_positions.append(self.joint_positions[joint])
            else:
                arm_positions.append(0.0)

        if len(arm_positions) == len(self.ARM_JOINTS):
            traj_msg = JointTrajectory()
            traj_msg.joint_names = self.ARM_JOINTS

            point = JointTrajectoryPoint()
            point.positions = arm_positions
            point.time_from_start = Duration(sec=0, nanosec=50000000)  # 50ms

            traj_msg.points = [point]
            self.arm_pub.publish(traj_msg)

        # Gripper position 발행
        gripper_positions = []
        for joint in self.GRIPPER_JOINTS:
            if joint in self.joint_positions:
                gripper_positions.append(self.joint_positions[joint])
            else:
                gripper_positions.append(0.0)

        if gripper_positions:
            gripper_msg = Float64MultiArray()
            gripper_msg.data = gripper_positions
            self.gripper_pub.publish(gripper_msg)

        self.count += 1
        if self.count % 500 == 0:  # 10초마다 로그
            self.get_logger().info(f'Published {self.count} updates to Gazebo')


def main():
    import argparse
    parser = argparse.ArgumentParser(description='Gazebo Bridge')
    parser.add_argument('--real-ns', default='dsr01', help='Real robot namespace')
    parser.add_argument('--gazebo-ns', default='gz', help='Gazebo robot namespace')
    args = parser.parse_args()

    rclpy.init()
    node = GazeboBridge(real_ns=args.real_ns, gazebo_ns=args.gazebo_ns)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        print('\nShutting down...')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: e0509_gripper_description/src/gazebo_bridge.cpp
/*
 * gazebo_bridge.cpp
 * Real robot → Gazebo synchronization bridge
 *
 * Subscribes to real robot joint_states and publishes to Gazebo controllers
 * for digital twin visualization at 50Hz.
 */

#include <map>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"
#include "trajectory_msgs/msg/joint_trajectory_point.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"

using namespace std::chrono_literals;

class GazeboBridge : public rclcpp::Node {
public:
    GazeboBridge(const std::string& real_ns = "dsr01", const std::string& gazebo_ns = "gz")
        : Node("gazebo_bridge"), real_ns_(real_ns), gazebo_ns_(gazebo_ns), count_(0)
    {
        // Subscribe to real robot joint_states
        std::string real_topic = "/" + real_ns + "/joint_states";
        real_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
            real_topic, 10,
            std::bind(&GazeboBridge::real_joint_state_callback, this, std::placeholders::_1));

        // Publish to Gazebo arm trajectory controller
        std::string arm_topic = "/" + gazebo_ns + "/joint_trajectory_controller/joint_trajectory";
        arm_pub_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(arm_topic, 10);

        // Publish to Gazebo gripper controller
        std::string gripper_topic = "/" + gazebo_ns + "/gripper_controller/commands";
        gripper_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(gripper_topic, 10);

        // 50Hz publish timer
        timer_ = this->create_wall_timer(20ms, std::bind(&GazeboBridge::publish_to_gazebo, this));

        RCLCPP_INFO(this->get_logger(), "============================================================");
        RCLCPP_INFO(this->get_logger(), "  Gazebo Digital Twin Bridge Started");
        RCLCPP_INFO(this->get_logger(), "============================================================");
        RCLCPP_INFO(this->get_logger(), "  Real robot namespace: %s", real_ns.c_str());
        RCLCPP_INFO(this->get_logger(), "  Gazebo namespace: %s", gazebo_ns.c_str());
        RCLCPP_INFO(this->get_logger(), "  Subscribed to: %s", real_topic.c_str());
        RCLCPP_INFO(this->get_logger(), "  Publishing to: %s", arm_topic.c_str());
        RCLCPP_INFO(this->get_logger(), "  Publishing to: %s", gripper_topic.c_str());
        RCLCPP_INFO(this->get_logger(), "============================================================");
    }

private:
    void real_joint_state_callback(const sensor_msgs::msg::JointState::SharedPtr msg) {
        for (size_t i = 0; i < msg->name.size() && i < msg->position.size(); ++i) {
            joint_positions_[msg->name[i]] = msg->position[i];
        }
    }

    void publish_to_gazebo() {
        if (joint_positions_.empty()) return;

        // Arm trajectory
        std::vector<double> arm_positions;
        for (const auto& joint : ARM_JOINTS) {
            auto it = joint_positions_.find(joint);
            arm_positions.push_back(it != joint_positions_.end() ? it->second : 0.0);
        }

        if (arm_positions.size() == ARM_JOINTS.size()) {
            auto traj_msg = trajectory_msgs::msg::JointTrajectory();
            traj_msg.joint_names = ARM_JOINTS;

            auto point = trajectory_msgs::msg::JointTrajectoryPoint();
            point.positions = arm_positions;
            point.time_from_start.sec = 0;
            point.time_from_start.nanosec = 50000000;  // 50ms

            traj_msg.points.push_back(point);
            arm_pub_->publish(traj_msg);
        }

        // Gripper positions
        std::vector<double> gripper_positions;
        for (const auto& joint : GRIPPER_JOINTS) {
            auto it = joint_positions_.find(joint);
            gripper_positions.push_back(it != joint_positions_.end() ? it->second : 0.0);
        }

        if (!gripper_positions.empty()) {
            auto gripper_msg = std_msgs::msg::Float64MultiArray();
            gripper_msg.data = gripper_positions;
            gripper_pub_->publish(gripper_msg);
        }

        ++count_;
        if (count_ % 500 == 0) {
            RCLCPP_INFO(this->get_logger(), "Published %d updates to Gazebo", count_);
        }
    }

    const std::vector<std::string> ARM_JOINTS = {
        "joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"
    };
    const std::vector<std::string> GRIPPER_JOINTS = {
        "gripper_rh_r1", "gripper_rh_r2", "gripper_rh_l1", "gripper_rh_l2"
    };

    std::string real_ns_;
    std::string gazebo_ns_;
    int count_;
    std::map<std::string, double> joint_positions_;

    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr real_sub_;
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr arm_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr gripper_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char** argv) {
    rclcpp::init(argc, argv);

    std::string real_ns = "dsr01";
    std::string gazebo_ns = "gz";

    for (int i = 1; i < argc; ++i) {
        std::string arg(argv[i]);
        if (arg == "--real-ns" && i + 1 < argc) {
            real_ns = argv[++i];
        } else if (arg == "--gazebo-ns" && i + 1 < argc) {
            gazebo_ns = argv[++i];
        }
    }

    auto node = std::make_shared<GazeboBridge>(real_ns, gazebo_ns);

    rclcpp::spin(node);
    node.reset();
    rclcpp::shutdown();
    return 0;
}
